lib/web_ui/lib/src/engine/text_editing/text_editing.dart

```diff
@@ -440,6 +440,211 @@ class AutofillInfo {
+      // a native composing menu.
+      final ui.TextRange replacementRange = ui.TextRange(start: newTextEditingDeltaState.deltaStart, end: newTextEditingDeltaState.deltaEnd);
+      final String textAfterDelta = _replace(
+          newTextEditingDeltaState.oldText, newTextEditingDeltaState.deltaText,
+          replacementRange);
+      final bool isDeltaVerified = textAfterDelta == newEditingState.text!;
+
+      if (!isDeltaVerified) {
+        // 1. Find all matches for deltaText.
+        // 2. Apply matches/replacement to oldText until oldText matches the
+        // new editing state's text value.
+        final bool isPeriodInsertion = newTextEditingDeltaState.deltaText.contains('.');
+        final RegExp deltaTextPattern = isPeriodInsertion?
+                                        RegExp(r'\' + newTextEditingDeltaState.deltaText + r'')
+                                            : RegExp(r'' + newTextEditingDeltaState.deltaText + r'');
+        for (final Match match in deltaTextPattern.allMatches(newEditingState.text!)) {
+          String textAfterMatch;
+          int actualEnd;
+          final bool isMatchWithinOldTextBounds = match.start >= 0 && match.end <= newTextEditingDeltaState.oldText.length;
+          if (!isMatchWithinOldTextBounds) {
+            actualEnd = match.start + newTextEditingDeltaState.deltaText.length - 1;
+            textAfterMatch = _replace(
+              newTextEditingDeltaState.oldText,
+              newTextEditingDeltaState.deltaText,
+              ui.TextRange(
+                start: match.start,
+                end: actualEnd,
+              ),
+            );
+          } else {
+            actualEnd = actualEnd = isPeriodInsertion? match.end - 1 : match.end;
+            textAfterMatch = _replace(
+              newTextEditingDeltaState.oldText,
+              newTextEditingDeltaState.deltaText,
+              ui.TextRange(
+                start: match.start,
+                end: actualEnd,
+              ),
+            );
+          }
+
+          if (textAfterMatch == newEditingState.text!) {
+            newTextEditingDeltaState.deltaStart = match.start;
+            newTextEditingDeltaState.deltaEnd = actualEnd;
+            break;
+          }
+        }
+      }
+    }
+
+    // Update selection of the delta using information from the new editing state.
+    newTextEditingDeltaState.baseOffset = newEditingState.baseOffset!;
+    newTextEditingDeltaState.extentOffset = newEditingState.extentOffset!;
+
+    return newTextEditingDeltaState;
+  }
+
+  /// The text before the text field was updated.
+  String oldText;
+
+  /// The text that is being inserted/replaced into the text field.
+  /// This will be an empty string for deletions and non text updates
+  /// such as selection updates.
+  String deltaText;
+
+  /// The position in the text field where the change begins.
+  int deltaStart;
+
+  /// The position in the text field where the change ends.
+  int deltaEnd;
```

Comment:
If empty strings and -1 are meaningful values for these fields, let's explain what they mean. Something tells me that the 3 fields above could be combined into a `TextEditingDelta(text, start, end)` struct and the default here expressed as `TextEditingDelta? delta;`. I may be missing something though.

Reply:
I think since the selection and composing region change when the text is mutated they should be together with `deltaStart`,`deltaEnd`, and `deltaText` in one class.